src/main/scala/nl/knaw/dans/easy/multideposit/Action.scala

```diff
@@ -15,32 +15,155 @@
+  }
+  private def logRollback(): Unit = {
+    logger.info(s"An error occurred. Rolling back action ${getClass.getSimpleName} ...")
+  }
+  private[Action] def innerCheckPreconditions: Try[Unit] = Try(logPreconditions()).flatMap(_ => checkPreconditions)
+  private[Action] def innerExecute(): Try[Unit] = Try(logExecute()).flatMap(_ => execute())
+  private[Action] def innerRollback(): Try[Unit] = Try(logRollback()).flatMap(_ => rollback())
+
+  /**
+   * Verifies whether all preconditions are met for this specific action.
+   *
+   * @return `Success` when all preconditions are met, `Failure` otherwise
+   */
+  protected def checkPreconditions: Try[Unit] = Success(())
+
+  /**
+   * Exectue the action.
+   *
+   * @return `Success` if the execution was successful, `Failure` otherwise
+   */
+  protected def execute(): Try[Unit]
 
   /**
-    * Verifies whether all preconditions are met for this specific action.
-    *
-    * @return `Success` when all preconditions are met, `Failure` otherwise
-    */
-  def checkPreconditions: Try[Unit] = Success(())
+   * Cleans up results of a previous call to run so that a new call to run will not fail because of those results.
+   *
+   * @return `Success` if the rollback was successful, `Failure` otherwise
+   */
+  protected def rollback(): Try[Unit] = Success(())
 
   /**
-    * Exectue the action.
-    *
-    * @return `Success` if the execution was successfull, `Failure` otherwise
-    */
-  def run(): Try[Unit]
+   * Run an action. First the precondition is checked. If it fails a `PreconditionsFailedException`
+   * with a report is returned. Else, if the precondition succeed, the action is executed.
+   * If this fails, the action is rolled back and a `ActionRunFailedException` with a report is returned.
+   * If the execution was successful, `Success` is returned
+   *
+   * @return `Success` if the full execution was successful, `Failure` otherwise
+   */
+  final def run: Try[Unit] = {
+    for {
+      _ <- innerCheckPreconditions.recoverWith {
+        case NonFatal(e) =>
+          Failure(PreconditionsFailedException(generateReport("Precondition failures:", e, "Due to these errors in the preconditions, nothing was done.")))
+      }
+      _ <- innerExecute().recoverWith {
+        case NonFatal(e) => List(Failure(e), innerRollback()).collectResults.recoverWith {
+          case e: CompositeException => Failure(ActionRunFailedException(generateReport("Errors in Multi-Deposit Instructions file:", e)))
+        }
+      }
+    } yield ()
+  }
+
+  private def generateReport(header: String = "", t: Throwable, footer: String = ""): String = {
+    val report = t match {
+      case ActionException(row, msg, _) => s" - row $row: $msg"
+      case CompositeException(ths) =>
+        ths.toSeq
+          .collect { case ActionException(row, msg, _) => s" - row $row: $msg" }
+          .mkString("\n")
+      case e => s" - unexpected error: ${e.getMessage}"
```

Comment:
Good point! Now it does...
I added a `cause` to the `ActionRunFailedException` and the `PreconditionsFailedException` and fixed this in the `run`. If you run the logger in debug mode, the stack traces will show up in the log.